File: codes/visual_basic_dotnet/array_every.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Function ArrayEveryV1(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Boolean
        ' JavaScript-like Array.every() function
        Dim IsConditionMatch As Boolean = false
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            IsConditionMatch = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            If (IsConditionMatch = false) Then Exit For
            ArrayItemIndex += 1
        Next
        Return IsConditionMatch
    End Function

    Function ArrayEveryV2(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Boolean
        ' JavaScript-like Array.every() function
        Dim IsConditionMatch As Boolean = false
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            IsConditionMatch = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            If (IsConditionMatch = false) Then Return IsConditionMatch
            ArrayItemIndex += 1
        Next
        Return IsConditionMatch
    End Function

    Function ArrayEveryV3(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Boolean
        ' JavaScript-like Array.every() function
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            Dim IsConditionMatch As Boolean = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            If (IsConditionMatch = false) Then Return false
            ArrayItemIndex += 1
        Next
        Return true
    End Function

    Function ArrayEveryV4(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Boolean
        ' JavaScript-like Array.every() function
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            If (CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) = false) Then Return false
            ArrayItemIndex += 1
        Next
        Return true
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.every() in Visual Basic (.NET) List")

        Dim Numbers As List(Of Object) = New List(Of Object) From {12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}")

        Dim IsAllNumberLessThan500 As Boolean
        Dim IsAllNumberMoreThan500 As Boolean

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV1""")

        IsAllNumberLessThan500 = ArrayEveryV1(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number < 500), Numbers)
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}")
        ' is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV1(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number > 500), Numbers)
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}")
        ' is all number > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV2""")

        IsAllNumberLessThan500 = ArrayEveryV2(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number < 500), Numbers)
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}")
        ' is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV2(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number > 500), Numbers)
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}")
        ' is all number > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV3""")

        IsAllNumberLessThan500 = ArrayEveryV3(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number < 500), Numbers)
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}")
        ' is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV3(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number > 500), Numbers)
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}")
        ' is all number > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV4""")

        IsAllNumberLessThan500 = ArrayEveryV4(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number < 500), Numbers)
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}")
        ' is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV4(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Number > 500), Numbers)
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}")
        ' is all number > 500: False

        Console.WriteLine("' using Visual Basic (.NET) Array.every() built-in method ""Array.All()""")

        IsAllNumberLessThan500 = Numbers.All(Function(ByVal Number As Integer) (Number < 500))
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}")
        ' is all number < 500: True

        IsAllNumberMoreThan500 = Numbers.All(Function(ByVal Number As Integer) (Number > 500))
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}")
        ' is all number > 500: False

        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.every() in Visual Basic (.NET) List of Dictionaries")

        Dim Products As List(Of Object) = New List(Of Object) From {
            New Dictionary(Of String, Object) From {
                {"code", "pasta"},
                {"price", 321}
            },
            New Dictionary(Of String, Object) From {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            New Dictionary(Of String, Object) From {
                {"code", "potato_chips"},
                {"price", 5}
            },
            New Dictionary(Of String, Object) From {
                {"code", "towel"},
                {"price", 499}
            }
        }
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty:=True)}")

        Dim IsAllProductPriceLessThan500 As Boolean
        Dim IsAllProductPriceMoreThan500 As Boolean

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV1""")

        IsAllProductPriceLessThan500 = ArrayEveryV1(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") < 500), Products)
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}")
        ' is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV1(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") > 500), Products)
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}")
        ' is all product price > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV2""")

        IsAllProductPriceLessThan500 = ArrayEveryV2(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") < 500), Products)
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}")
        ' is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV2(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") > 500), Products)
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}")
        ' is all product price > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV3""")

        IsAllProductPriceLessThan500 = ArrayEveryV3(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") < 500), Products)
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}")
        ' is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV3(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") > 500), Products)
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}")
        ' is all product price > 500: False

        Console.WriteLine("' using JavaScript-like Array.every() function ""ArrayEveryV4""")

        IsAllProductPriceLessThan500 = ArrayEveryV4(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") < 500), Products)
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}")
        ' is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV4(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("price") > 500), Products)
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}")
        ' is all product price > 500: False

        Console.WriteLine("' using Visual Basic (.NET) Array.every() built-in method ""Array.All()""")

        IsAllProductPriceLessThan500 = Products.All(Function(ByVal Product As Object) (Product("price") < 500))
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}")
        ' is all product price < 500: True

        IsAllProductPriceMoreThan500 = Products.All(Function(ByVal Product As Object) (Product("price") > 500))
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}")
        ' is all product price > 500: False
    End Sub
End Module

File: codes/javascript/src/array_every.js
const jsType = {
    "Null": "Null",
    "Boolean": "Boolean",
    "String": "String",
    "Numeric": "Numeric",
    "Object": "Object",
    "Array": "Array",
    "Function": "Function"
};

const isNull = (anything) => (((Object.prototype.toString.call(anything) === "[object Null]") && (anything === null)) || ((Object.prototype.toString.call(anything) === "[object Undefined]") && (anything === undefined)));

const isBoolean = (anything) => ((Object.prototype.toString.call(anything) === "[object Boolean]") && ((anything === true) || (anything === false)));

const isString = (anything) => (Object.prototype.toString.call(anything) === "[object String]");

const isNumeric = (anything) => ((Object.prototype.toString.call(anything) === "[object Number]") && (Number.isNaN(anything) === false) && (Number.isFinite(anything) === true));

const isObject = (anything) => (Object.prototype.toString.call(anything) === "[object Object]");

const isArray = (anything) => ((Object.prototype.toString.call(anything) === "[object Array]") && (Array.isArray(anything) === true));

const isFunction = (anything) => (Object.prototype.toString.call(anything) === '[object Function]');

const getType = (anything) => {
    if (isNull(anything) === true) return jsType.Null;
    if (isBoolean(anything) === true) return jsType.Boolean;
    if (isString(anything) === true) return jsType.String;
    if (isNumeric(anything) === true) return jsType.Numeric;
    if (isObject(anything) === true) return jsType.Object;
    if (isArray(anything) === true) return jsType.Array;
    if (isFunction(anything) === true) return jsType.Function;
    return Object.prototype.toString.call(anything);
};

const jsonStringify = (anything, { pretty = false } = {}) => {
    // custom JSON.stringify() function jsonStringifyV3
    const indent = " ".repeat(4);
    let indentLevel = 0;
    const jsonStringifyInner = (anythingInner) => {
        if (getType(anythingInner) === jsType.Null) return "null";
        if (getType(anythingInner) === jsType.String) return `"${anythingInner}"`;
        if ((getType(anythingInner) === jsType.Numeric) || (getType(anythingInner) === jsType.Boolean)) return `${anythingInner}`;
        if (getType(anythingInner) === jsType.Object) {
            if (Object.keys(anythingInner).length === 0) return "{}";
            indentLevel += 1;
            let result = ((pretty === true) ? (`{\n${indent.repeat(indentLevel)}`) : "{ ");
            Object.entries(anythingInner).forEach(([objectKey, objectValue], objectEntryIndex) => {
                result = `${result}"${objectKey}": ${jsonStringifyInner(objectValue)}`;
                if ((objectEntryIndex + 1) !== Object.keys(anythingInner).length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}}`) : " }")}`;
            return result;
        }
        if (getType(anythingInner) === jsType.Array) {
            if (anythingInner.length === 0) return "[]";
            indentLevel += 1;
            let result = ((pretty === true) ? (`[\n${indent.repeat(indentLevel)}`) : "[");
            anythingInner.forEach((arrayItem, arrayItemIndex) => {
                result = `${result}${jsonStringifyInner(arrayItem)}`;
                if ((arrayItemIndex + 1) !== anythingInner.length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}]`) : "]")}`;
            return result;
        }
        if (getType(anythingInner) === jsType.Function) return '"[object Function]"';
        return `"${getType(anything)}"`;
    };
    return jsonStringifyInner(anything);
};

console.log("\n// Array.every() in JavaScript Array");

const numbers = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
console.log(`numbers: ${jsonStringify(numbers)}`);

const isAllNumberLessThan500 = numbers.every((number) => (number < 500));
console.log(`is all number < 500: ${isAllNumberLessThan500}`);
// is all number < 500: true

const isAllNumberMoreThan500 = numbers.every((number) => (number > 500));
console.log(`is all number > 500: ${isAllNumberMoreThan500}`);
// is all number > 500: false

console.log("\n// Array.every() in JavaScript Array of Objects");

const products = [
    {
        "code": "pasta",
        "price": 321
    },
    {
        "code": "bubble_gum",
        "price": 233
    },
    {
        "code": "potato_chips",
        "price": 5
    },
    {
        "code": "towel",
        "price": 499
    }
];
console.log(`products: ${jsonStringify(products, { pretty: true })}`);

const isAllProductPriceLessThan500 = products.every((product) => (product?.price < 500));
console.log(`is all product price < 500: ${isAllProductPriceLessThan500}`);
// is all product price < 500: true

const isAllProductPriceMoreThan500 = products.every((product) => (product?.price > 500));
console.log(`is all product price > 500: ${isAllProductPriceMoreThan500}`);
// is all product price > 500: false
